File: src/precalculus_polar_curves_easy.py
"""
precalculus - polar_curves (easy)
"""

from problem_utils import *

THETA_OPTIONS = [0, pi/6, pi/4, pi/3, pi/2, 2*pi/3, 3*pi/4, 5*pi/6,
                 pi, 7*pi/6, 5*pi/4, 4*pi/3, 3*pi/2, 5*pi/3, 7*pi/4, 11*pi/6]


def generate():
    problem_type = randint(1, 6)

    if problem_type == 1:
        # Convert polar point to Cartesian — wide r and theta range
        r_val = randint(1, 12)
        theta_val = choice(THETA_OPTIONS)

        x_coord = simplify(r_val * cos(theta_val))
        y_coord = simplify(r_val * sin(theta_val))

        return problem(
            question=f"Convert the polar point $\\left({r_val},\\, {latex(theta_val)}\\right)$ to Cartesian coordinates $(x, y)$.",
            answer=(x_coord, y_coord),
            difficulty=(1000, 1200),
            topic="precalculus/polar_curves",
            solution=steps(
                f"$x = r\\cos\\theta = {r_val}\\cos\\!\\left({latex(theta_val)}\\right) = {latex(x_coord)}$",
                f"$y = r\\sin\\theta = {r_val}\\sin\\!\\left({latex(theta_val)}\\right) = {latex(y_coord)}$",
                f"Answer: $\\left({latex(x_coord)},\\, {latex(y_coord)}\\right)$"
            ),
            calculator="scientific"
        )

    elif problem_type == 2:
        # Evaluate r for a simple polar curve at a given theta
        a_val = randint(1, 10)
        curve_type = choice(['constant', 'cosine', 'sine', 'linear_theta'])

        if curve_type == 'constant':
            theta_val = choice(THETA_OPTIONS)
            ans = a_val
            return problem(
                question=f"For the polar curve $r = {a_val}$, find $r$ when $\\theta = {latex(theta_val)}$.",
                answer=ans,
                difficulty=(1000, 1100),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"$r = {a_val}$ is constant for all $\\theta$",
                    f"$r = {a_val}$"
                )
            )

        elif curve_type == 'cosine':
            theta_val = choice([0, pi/6, pi/4, pi/3, pi/2, 2*pi/3, 3*pi/4, pi])
            ans = simplify(a_val * cos(theta_val))
            return problem(
                question=f"For the polar curve $r = {a_val}\\cos(\\theta)$, find $r$ when $\\theta = {latex(theta_val)}$.",
                answer=ans,
                difficulty=(1100, 1200),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"Substitute: $r = {a_val}\\cos\\!\\left({latex(theta_val)}\\right) = {a_val} \\cdot {latex(cos(theta_val))} = {latex(ans)}$"
                ),
                calculator="scientific"
            )

        elif curve_type == 'sine':
            theta_val = choice([0, pi/6, pi/4, pi/3, pi/2, 2*pi/3, 3*pi/4, pi])
            ans = simplify(a_val * sin(theta_val))
            return problem(
                question=f"For the polar curve $r = {a_val}\\sin(\\theta)$, find $r$ when $\\theta = {latex(theta_val)}$.",
                answer=ans,
                difficulty=(1100, 1200),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"Substitute: $r = {a_val}\\sin\\!\\left({latex(theta_val)}\\right) = {a_val} \\cdot {latex(sin(theta_val))} = {latex(ans)}$"
                ),
                calculator="scientific"
            )

        else:  # r = a*theta
            theta_val = choice([pi/6, pi/4, pi/3, pi/2, pi])
            ans = simplify(a_val * theta_val)
            return problem(
                question=f"For the spiral $r = {a_val}\\theta$, find $r$ when $\\theta = {latex(theta_val)}$.",
                answer=ans,
                difficulty=(1100, 1200),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"Substitute: $r = {a_val} \\cdot {latex(theta_val)} = {latex(ans)}$"
                ),
                calculator="scientific"
            )

    elif problem_type == 3:
        # Convert Cartesian to polar: find r
        x_val = choice([0, 1, 2, 3, 4, 5, 6, 7, 8])
        y_val = choice([0, 1, 2, 3, 4, 5, 6, 7, 8])
        while x_val == 0 and y_val == 0:
            x_val = randint(1, 8)
        r_sq = x_val**2 + y_val**2
        ans = sqrt(r_sq)

        return problem(
            question=f"Find $r$ for the Cartesian point $({x_val},\\, {y_val})$.",
            answer=ans,
            difficulty=(1100, 1300),
            topic="precalculus/polar_curves",
            solution=steps(
                f"$r = \\sqrt{{x^2 + y^2}} = \\sqrt{{{x_val}^2 + {y_val}^2}} = \\sqrt{{{r_sq}}} = {latex(ans)}$"
            ),
            calculator="scientific"
        )

    elif problem_type == 4:
        # Identify curve type from polar equation
        curve_choice = randint(1, 5)

        if curve_choice == 1:
            a_val = randint(1, 8)
            return problem(
                question=f"What shape is the polar curve $r = {a_val}$?",
                answer="circle",
                difficulty=(1000, 1100),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"$r = {a_val}$ means constant distance from the origin — a circle of radius ${a_val}$"
                )
            )
        elif curve_choice == 2:
            theta_val = choice([pi/6, pi/4, pi/3, pi/2, 2*pi/3, 3*pi/4])
            return problem(
                question=f"What shape is the polar curve $\\theta = {latex(theta_val)}$?",
                answer="line",
                difficulty=(1050, 1150),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"Constant angle $\\theta = {latex(theta_val)}$ describes a ray (line) through the origin"
                )
            )
        elif curve_choice == 3:
            a_val = randint(1, 5)
            return problem(
                question=f"What is the name of the curve $r = {a_val}(1 + \\cos\\theta)$?",
                answer="cardioid",
                difficulty=(1200, 1300),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"The form $r = a(1 + \\cos\\theta)$ is a cardioid (heart-shaped curve)"
                )
            )
        elif curve_choice == 4:
            n = choice([2, 3, 4, 5])
            a_val = randint(1, 5)
            return problem(
                question=f"How many petals does the rose $r = {a_val}\\cos({n}\\theta)$ have?",
                answer=n if n % 2 == 1 else 2*n,
                difficulty=(1150, 1300),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"For $r = a\\cos(n\\theta)$: odd $n$ gives $n$ petals; even $n$ gives $2n$ petals",
                    f"$n = {n}$ is {'odd' if n % 2 == 1 else 'even'}: {n if n % 2 == 1 else 2*n} petals"
                )
            )
        else:
            a_val = randint(1, 5)
            return problem(
                question=f"Is the curve $r = {a_val}\\cos\\theta$ a circle or a cardioid?",
                answer="circle",
                difficulty=(1150, 1250),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"$r = {a_val}\\cos\\theta$ converts to $(x - \\frac{{{a_val}}}{{2}})^2 + y^2 = (\\frac{{{a_val}}}{{2}})^2$, a circle"
                )
            )

    elif problem_type == 5:
        # Find smallest positive theta where r = 0 for r = a*cos or a*sin
        a_val = randint(1, 8)
        trig_choice = choice(['cos', 'sin', 'cos2', 'sin2'])

        if trig_choice == 'cos':
            return problem(
                question=f"Find the smallest positive $\\theta$ where $r = 0$ for $r = {a_val}\\cos(\\theta)$.",
                answer=pi/2,
                difficulty=(1200, 1300),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"Set $r = 0$: $\\cos\\theta = 0$",
                    f"Smallest positive solution: $\\theta = \\dfrac{{\\pi}}{{2}}$"
                ),
                calculator="scientific"
            )
        elif trig_choice == 'sin':
            return problem(
                question=f"Find the smallest positive $\\theta$ where $r = 0$ for $r = {a_val}\\sin(\\theta)$.",
                answer=pi,
                difficulty=(1200, 1300),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"Set $r = 0$: $\\sin\\theta = 0$",
                    f"Smallest positive solution: $\\theta = \\pi$"
                ),
                calculator="scientific"
            )
        elif trig_choice == 'cos2':
            return problem(
                question=f"Find the smallest positive $\\theta$ where $r = 0$ for $r = {a_val}\\cos(2\\theta)$.",
                answer=pi/4,
                difficulty=(1200, 1300),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"Set $r = 0$: $\\cos(2\\theta) = 0 \\Rightarrow 2\\theta = \\dfrac{{\\pi}}{{2}}$",
                    f"Smallest positive: $\\theta = \\dfrac{{\\pi}}{{4}}$"
                ),
                calculator="scientific"
            )
        else:  # sin2
            return problem(
                question=f"Find the smallest positive $\\theta$ where $r = 0$ for $r = {a_val}\\sin(2\\theta)$.",
                answer=pi/2,
                difficulty=(1200, 1300),
                topic="precalculus/polar_curves",
                solution=steps(
                    f"Set $r = 0$: $\\sin(2\\theta) = 0 \\Rightarrow 2\\theta = \\pi$",
                    f"Smallest positive: $\\theta = \\dfrac{{\\pi}}{{2}}$"
                ),
                calculator="scientific"
            )

    else:
        # Maximum value of r for simple polar curves
        a_val = randint(1, 8)
        b_val = randint(1, 8)
        limacon_type = choice(['cos', 'sin'])
        trig_str = '\\cos' if limacon_type == 'cos' else '\\sin'

        ans = a_val + b_val

        return problem(
            question=f"Find the maximum value of $r$ for the curve $r = {a_val} + {b_val}{trig_str}(\\theta)$.",
            answer=ans,
            difficulty=(1100, 1250),
            topic="precalculus/polar_curves",
            solution=steps(
                f"Maximum of ${trig_str}(\\theta)$ is $1$",
                f"Maximum $r = {a_val} + {b_val} \\cdot 1 = {ans}$"
            )
        )

emit(generate())

File: src/precalculus_polar_curves_easy.jl
# precalculus - polar_curves (easy)
# Generated: 2026-03-08T20:33:51.984062

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script r θ begin
    set_topic!("precalculus/polar_curves")
    
    prob_type = choice(1:5)
    
    if prob_type == 1
        # Convert simple polar point to Cartesian
        r_val = randint(2, 12)
        θ_val = choice([0, π/6, π/4, π/3, π/2, 2π/3, 3π/4, 5π/6, π, 7π/6, 5π/4, 4π/3, 3π/2, 5π/3, 7π/4, 11π/6])
        
        x_ans = r_val * cos(θ_val)
        y_ans = r_val * sin(θ_val)
        
        θ_display = if θ_val == 0
            "0"
        elseif θ_val == π
            "\\pi"
        elseif θ_val == π/2
            "\\frac{\\pi}{2}"
        elseif θ_val == 3π/2
            "\\frac{3\\pi}{2}"
        elseif θ_val == π/3
            "\\frac{\\pi}{3}"
        elseif θ_val == π/4
            "\\frac{\\pi}{4}"
        elseif θ_val == π/6
            "\\frac{\\pi}{6}"
        elseif θ_val == 2π/3
            "\\frac{2\\pi}{3}"
        elseif θ_val == 3π/4
            "\\frac{3\\pi}{4}"
        elseif θ_val == 5π/6
            "\\frac{5\\pi}{6}"
        elseif θ_val == 4π/3
            "\\frac{4\\pi}{3}"
        elseif θ_val == 5π/3
            "\\frac{5\\pi}{3}"
        elseif θ_val == 7π/6
            "\\frac{7\\pi}{6}"
        elseif θ_val == 5π/4
            "\\frac{5\\pi}{4}"
        elseif θ_val == 7π/4
            "\\frac{7\\pi}{4}"
        else
            "\\frac{11\\pi}{6}"
        end
        
        problem(
            question="Convert the polar point \\(($(r_val), $(θ_display))\\) to Cartesian coordinates.",
            answer=fmt_tuple([x_ans, y_ans]),
            difficulty=(700, 850),
            answer_type="point",
            solution=steps(
                "Use the conversion formulas: \$x = r\\cos(\\theta)\$ and \$y = r\\sin(\\theta)\$",
                "\$x = $(r_val)\\cos($(θ_display)) = $(x_ans)\$",
                "\$y = $(r_val)\\sin($(θ_display)) = $(y_ans)\$",
                "Answer: \$($(x_ans), $(y_ans))\$"
            ),
            time=90,
        )
        
    elseif prob_type == 2
        # Convert Cartesian point to polar (positive r)
        x_val = choice([-6, -4, -3, -2, -1, 1, 2, 3, 4, 6])
        y_val = choice([-6, -4, -3, -2, -1, 1, 2, 3, 4, 6])
        
        r_ans = sqrt(x_val^2 + y_val^2)
        θ_ans = atan(y_val, x_val)
        if θ_ans < 0
            θ_ans += 2π
        end
        
        problem(
            question="Convert the Cartesian point \\(($(x_val), $(y_val))\\) to polar coordinates \\((r, \\theta)\\) where \\(r > 0\\) and \\(0 \\le \\theta < 2\\pi\\). Express \\(\\theta\\) in radians.",
            answer=fmt_tuple([r_ans, θ_ans]),
            difficulty=(850, 1000),
            answer_type="point",
            solution=steps(
                "Use \$r = \\sqrt{x^2 + y^2}\$ and \$\\theta = \\arctan(y/x)\$ (adjusted for quadrant)",
                "\$r = \\sqrt{$(x_val)^2 + $(y_val)^2} = \\sqrt{$(x_val^2 + y_val^2)} = $(r_ans)\$",
                "\$\\theta = \\arctan($(y_val)/$(x_val)) = $(θ_ans)\$ radians",
                "Answer: \$($(r_ans), $(θ_ans))\$"
            ),
            calculator="scientific",
            time=120,
        )
        
    elseif prob_type == 3
        # Identify polar curve type
        curve_choice = choice(1:4)
        
        if curve_choice == 1
            a = randint(2, 8)
            equation = "r = $(a)"
            answer = "circle centered at origin with radius $(a)"
            curve_type = "Circle"
        elseif curve_choice == 2
            a = randint(2, 8)
            equation = "r = $(a)\\cos(\\theta)"
            answer = "circle"
            curve_type = "Circle (shifted)"
        elseif curve_choice == 3
            a = randint(2, 8)
            n = choice([3, 4, 5, 6])
            equation = "r = $(a)\\sin($(n)\\theta)"
            answer = "rose curve with $(n) petals"
            curve_type = "Rose"
        else
            a = randint(2, 8)
            equation = "r = $(a) + $(a)\\cos(\\theta)"
            answer = "cardioid"
            curve_type = "Cardioid"
        end
        
        problem(
            question="Identify the type of curve represented by the polar equation \$$(equation)\$.",
            answer=answer,
            difficulty=(750, 900),
            answer_type="text",
            solution=steps(
                "Given: \$$(equation)\$",
                "Recognize the standard form",
                "This is a $(curve_type)",
                "Answer: $(answer)"
            ),
            time=60,
        )
        
    elseif prob_type == 4
        # Find r at specific θ
        a = randint(2, 8)
        form_choice = choice(1:3)
        
        if form_choice == 1
            θ_val = choice([0, π/4, π/2, 3π/4, π])
            θ_display = θ_val == 0 ? "0" : θ_val == π ? "\\pi" : θ_val == π/2 ? "\\frac{\\pi}{2}" : θ_val == π/4 ? "\\frac{\\pi}{4}" : "\\frac{3\\pi}{4}"
            r_ans = a + a*cos(θ_val)
            equation = "r = $(a) + $(a)\\cos(\\theta)"
        elseif form_choice == 2
            θ_val = choice([0, π/6, π/4, π/3, π/2])
            θ_display = θ_val == 0 ? "0" : θ_val == π/2 ? "\\frac{\\pi}{2}" : θ_val == π/3 ? "\\frac{\\pi}{3}" : θ_val == π/4 ? "\\frac{\\pi}{4}" : "\\frac{\\pi}{6}"
            r_ans = a*cos(θ_val)
            equation = "r = $(a)\\cos(\\theta)"
        else
            θ_val = choice([0, π/6, π/4, π/3, π/2])
            θ_display = θ_val == 0 ? "0" : θ_val == π/2 ? "\\frac{\\pi}{2}" : θ_val == π/3 ? "\\frac{\\pi}{3}" : θ_val == π/4 ? "\\frac{\\pi}{4}" : "\\frac{\\pi}{6}"
            r_ans = a*sin(θ_val)
            equation = "r = $(a)\\sin(\\theta)"
        end
        
        problem(
            question="Find the value of \\(r\\) when \\(\\theta = $(θ_display)\\) for the polar curve \$$(equation)\$.",
            answer=r_ans,
            difficulty=(700, 850),
            solution=steps(
                "Given: \$$(equation)\$",
                "Substitute \$\\theta = $(θ_display)\$",
                "Calculate the value",
                sol("Answer", r_ans)
            ),
            time=75,
        )
        
    else
        # Convert polar equation to Cartesian
        eq_choice = choice(1:3)
        
        if eq_choice == 1
            a = randint(2, 10)
            polar_eq = "r = $(a)"
            cartesian_eq = "x^2 + y^2 = $(a^2)"
            answer = cartesian_eq
        elseif eq_choice == 2
            a = randint(2, 8)
            polar_eq = "r = $(a)\\cos(\\theta)"
            cartesian_eq = "x^2 + y^2 = $(a)x"
            answer = cartesian_eq
        else
            a = randint(2, 8)
            polar_eq = "r = $(a)\\sin(\\theta)"
            cartesian_eq = "x^2 + y^2 = $(a)y"
            answer = cartesian_eq
        end
        
        problem(
            question="Convert the polar equation \$$(polar_eq)\$ to Cartesian form.",
            answer=answer,
            difficulty=(900, 1100),
            answer_type="equation",
            solution=steps(
                "Given: \$$(polar_eq)\$",
                "Use \$r^2 = x^2 + y^2\$, \$x = r\\cos(\\theta)\$, \$y = r\\sin(\\theta)\$",
                "Multiply through and substitute",
                "Answer: \$$(cartesian_eq)\$"
            ),
            time=120,
        )
    end
end